### PR TITLE
Switch to cordova-plugin-advanced-http for update downloads

### PR DESCRIPTION
**Pull request by ermik**
WIP
- Needs tester 
- Needs review

Upd: July 15th, 2019 — this is back on the table for me; apologies for staleness.
Upd: November 21st, 2019 — resolved conflicts and fixed some mistakes.
Upd: April 16th, 2020 — @Krasavinigor jumped on to push this along.
Upd: June 24th, 2020 — I [found the issue](https://github.com/microsoft/cordova-plugin-code-push/pull/513#issuecomment-617834122) that was highlighted by Igor back in April, but haven't found a solution. I expect, now that's `cordova-ios@6.0` is out this will become a P1 on my radar, if somebody else doesn't resolve it sooner.
Upd: July 27th, 2020 — #632 is tracking fixes to test suite which should be implemented before this major dependency change is merged in.

This PR follows through on the #384 by removing dependency on `FileTransfer` plugin which has been deprecated (apache/cordova-plugin-file-transfer@265b63faa13792632d5953a194c6b6183b9a8e23). It also implements fixes to the type checker errors highlighted during development.

- [x] Find a suitable replacement for `FileTransfer` — originally we went with XHRv2 which is cited as the migration strategy following `FileTransfer` deprecation, but it has raised issues with CORS (especially in a [strict WKWebView networking environment](https://blog.ionicframework.com/wkwebview-for-all-a-new-webview-for-ionic/) on iOS). Native alternative is `cordova-plugin-http` which has been superseded by `-advanced-http`
- [x] (partially) Maintain type-safety — unfortunately `cordova-plugin-advanced-http` is written using common js and lacks typings. Additionally it writes to a non-standard cordova global `cordova.plugin.http` (`plugin`-singular) and we have to amend the definition coming from `@types/cordova`.
- [ ] Verify test suite passes — currently tracked in #632 
- [x] Verify in production environment

This PR now also replaces regular request utility, since plain XHR in iOS 13 may be failing.

### CODE REVIEW

**Comment by ermik** [comment]
This is designed to follow #384, with deprecation of`FileTransfer` still posing an issue.

There's a bit of a mess with getting the branch back on track, please let me know if you'd like me to rebase or something of that sort.

**Comment by ermik** [comment]
@dlebu @sergey-akhalkov is `downloadUrl` sufficient or would this approach have to also set headers on XHR akin ones seen in `HttpRequester` implementation?

**Comment by ermik** [comment]
I've updated the first comment to reflect the current state of the PR.

**Comment by jacobg** [comment]
Thanks for this @ermik!

Is this work ready to be merged into master?

@alexandergoncharov

**Comment by ermik** [comment]
@jacobg I wish, but not yet. We're still waiting to know what exactly are the requirements for initiating the download  — there are no documentation of CodePush API that I could find, particularly how the `downloadUrl` is created and any headers that need to exist on the request when we are attempting to download the archive. Move to `advanced-http` has proven to be the right way forward, as described in the top comment, but I'd like to double check everything before we merge this.

Is there anyone on CodePush team that can speak to how download is/was set up and executed so we can make sure nothing will go wrong with new code?

File: www/remotePackage.ts [hunk 1]
```diff
@@ -59,41 +77,30 @@ class RemotePackage extends Package implements IRemotePackage {
...
      * @param abortSuccess Optional callback invoked if the abort operation succeeded.
      * @param abortError Optional callback invoked in case of an error.
      */
     public abortDownload(abortSuccess?: SuccessCallback<void>, abortError?: ErrorCallback): void {
         try {
-            if (this.currentFileTransfer) {
-                this.currentFileTransfer.abort();
-            
+            if (this.isDownloading) {
+                this.isDownloading = false;
```
Comment (ermik) [inline www/remotePackage.ts]: This is questionable, but `cordova-plugin-advanced-http` does not implement `Canecelable` / has no abort triggers I could track down.

**Comment by alexandergoncharov** [comment]
Hi @ermik 
Sorry for delay! Thank you for contributing. You've done good and useful job =)

On the [your question of the request headers](https://github.com/microsoft/cordova-plugin-code-push/pull/513#issuecomment-476817218): there are no requirements for explicit specifying headers. `downloadUrl` should be sufficient for downloading. 
I tested your PR. Downloading and updating work fine for me in IOS.
But in Android it was thrown exception `java.io.FileNotFoundException (No such file or directory)`. Did you face this?

Could you please clarify your status on this PR and what can I do to help you?

**Comment by Nashorn** [comment]
What ever happened to this PR? Will this solve the issues some people see right after installing code-push plugin for android platform? A white screen of death with "deviceready" never firing? Something to do with the FileTransfer?

**Comment by ermik** [comment]
Sorry for losing track of this. @alexandergoncharov thank you for clarifying the request policy. No I did not face this issue on Android and will investigate this coming week. @Nashorn deviceready is a Cordova primitive and any issue with event firing should be investigated with them. There is a possibility of update failing which results in a white screen (the app code doesn't load after it has been replaced in the bundle) but I can't comment further on what to do to avoid that.

**Comment by Nashorn** [comment]
> @Nashorn deviceready is a Cordova primitive and any issue with event firing should be investigated with them. 

Thanks for replying @ermik. I've attempted the code-push install on several different machines, 3 macs and 1 windows and each time, as soon as the cordova plugin is installed, 'deviceready' never fires. I've mucked with removing platforms and re-adding, then reinstalling all plugins, it never works right. BUT, it does work with a sample cordova project that has no plugins except the white-list plugin.

Important part here is, if i remove code-push from cordova, the app runs fine, nothing else changed. After several months and multiple machine installs, and multiple configurations of plugin versions, cordova versions (tried 7, 8 and 9.0 btw), it never worked.

**Comment by ermik** [comment]
@Nashorn you should definitely create an issue. A small repo reproducing this behavior would help, as you might imagine. If you can't do that, you should detail the steps that produce the error (re-trace your steps, in a manner of speaking). 

I am concerned with discussing your issue in this PR as it might not get the exposure — and therefore — the input of the team at-large.

File: typings/codePush.d.ts [hunk 2]
```diff
@@ -7,7 +7,15 @@
...
-        GET, HEAD, POST, PUT, DELETE, TRACE, OPTIONS, CONNECT, PATCH
+        GET = "GET",
+        HEAD = "HEAD",
+        POST = "POST",
+        PUT = "PUT",
+        DELETE = "DELETE",
+        TRACE = "TRACE",
+        OPTIONS = "OPTIONS",
+        CONNECT = "CONNECT",
+        PATCH = "PATCH",
```
Comment (ermik) [inline typings/codePush.d.ts]: Is there a typescript version limitation that would prevent string enums from used in this project? I can remove this before merge if that's the case.

**Comment by Nashorn** [comment]
> @Nashorn you should definitely create an issue. A small repo reproducing this behavior would help, as you might imagine. If you can't do that, you should detail the steps that produce the error (re-trace your steps, in a manner of speaking).
> 
> I am concerned with discussing your issue in this PR as it might not get the exposure — and therefore — the input of the team at-large.

I'll definitely open an issue with a sample project

File: typings/codePush.d.ts (hunk 2 above)
Comment (ermik) [reply typings/codePush.d.ts]: @alexandergoncharov  I guess since 1bee4b1b73ec47b0429e24a99a56810d290c80ab is a thing, this should be fine.

**Comment by jacobg** [comment]
Is this PR ready to merge?

**Comment by ermik** [comment]
I couldn't find an issue with android, this branch has been tested with android devices.

If @alexandergoncharov could track down what tests failed for him I can look into this more. Regardless, let's put this on the september roadmap so it has some sort of deadline.

**Comment by EvanWieland** [comment]
Is this going to be merged soon?

**Comment by ermik** [comment]
@EvanWieland  I've been super flaky on this one, so don't hold your breath. Will push some updates right now, but this PR needs someone who can report which tests broke (I tried and failed to find the time to make gulp operate locally).

**Comment by ermik** [comment]
@alexandergoncharov is there a way we can set up GitHub Actions for this repo to run tests in the cloud — if I get the output I can make sure everything passes?

**Comment by jacobg** [comment]
It's very sad that this pull request has been wasting away for a year.

**Comment by ermik** [comment]
I'd love some help getting this through appropriate tests/checks.

**Comment by Krasavinigor** [comment]
Hello @ermik!
Sorry for delay! Thank you for contributing.
I tested this with both platforms, ios works without issues, but on Android I got error:

```
2020-04-15 08:50:49.342 13930-13930/com.example.tpfa D/SystemWebChromeClient: file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/codePushUtil.js: Line 36 : [CodePush] Downloading update

2020-04-15 08:50:49.342 13930-13930/com.example.tpfa I/chromium: [INFO:CONSOLE(36)] "[CodePush] Downloading update", source: file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/codePushUtil.js (36)

2020-04-15 08:50:50.561 13930-13930/com.example.tpfa D/SystemWebChromeClient: file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/codePushUtil.js: Line 41 : [CodePush]  Could not access local package. Stage:undefinedError code: undefined. StackTrace: Error: Could not access local package. Stage:undefinedError code: undefined 
```
Did you face this?

**Comment by ermik** [comment]
@Krasavinigor I've not seen it, except recently for some of many Android devices. The error is obviously coming from system code, not due to any changes in this particular PR. 

Please share the names of the tests that are failing. One of the issues I faced is that this repo does not make use of GitHub Actions or other CI tools to perform automated testing in a valid environment. I have a hard time tracking down a test case that I can debug.

**Comment by Krasavinigor** [comment]
@ermik It's not a test's logs. I just added plugin with your changes and my app couldn't install the update.
Steps to reproduced: 
* Create a simple Cordova project
* Add android platform
* Add this plugin and set up for an android platform

After downloading the update, I've got this error: https://github.com/microsoft/cordova-plugin-code-push/pull/513#issuecomment-613891164

**Comment by ermik** [comment]
@Krasavinigor whelp, I guess we should really add a test to catch that! :D

**Comment by ermik** [comment]
Here's my current findings:

logcat stack trace tells us that problem is coming from `advanced-http` Java runtime:

```cs
[W Cordova-Plugin-HTTP tag is omitted]
Generic request error
com.silkimen.http.HttpRequest$HttpRequestException: java.io.FileNotFoundException: /data/user/0/com.example.app/files/codepush/download/update.zip: open failed: ENOENT (No such file or directory)
	at com.silkimen.http.HttpRequest.receive(HttpRequest.java:1836)
	at com.silkimen.cordovahttp.CordovaHttpDownload.processResponse(CordovaHttpDownload.java:36)
	at com.silkimen.cordovahttp.CordovaHttpBase.run(CordovaHttpBase.java:78)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.io.FileNotFoundException: /data/user/0/com.example.app/files/codepush/download/update.zip: open failed: ENOENT (No such file or directory)
	at libcore.io.IoBridge.open(IoBridge.java:452)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:87)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:72)
	at com.silkimen.http.HttpRequest.receive(HttpRequest.java:1834)
	... 5 more
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
	at libcore.io.Posix.open(Native Method)
	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
	at libcore.io.IoBridge.open(IoBridge.java:438)
	... 8 more
```

Exception from the core runtime is raised first in [`CordovaHttpBase.receive()` function](https://github.com/silkimen/cordova-plugin-advanced-http/blob/b25b7db4befdf612fa0f9ad72c7862300a5db634/src/android/com/silkimen/http/HttpRequest.java#L1836):
```java
    try {
      output = new BufferedOutputStream(new FileOutputStream(file), bufferSize);
    } catch (FileNotFoundException e) {
      throw new HttpRequestException(e);
    }
```

obviously, `FileNotFoundException` is an expected condition since it's being handled in a HTTP Request class.

This exception is then seen upstream, in context of [`CordovaHttpDownload.processResponse()` function](https://github.com/silkimen/cordova-plugin-advanced-http/blob/b25b7db4befdf612fa0f9ad72c7862300a5db634/src/android/com/silkimen/cordovahttp/CordovaHttpDownload.java#L36): 
```java
    if (request.code() >= 200 && request.code() < 300) {
      File file = new File(new URI(this.filePath));
      JSONObject fileEntry = FileUtils.getFilePlugin().getEntryForFile(file);

      request.receive(file);
      response.setFileEntry(fileEntry);
    } else {
      response.setErrorMessage("There was an error downloading the file");
    }
```

where we can see that request has succeeded within the `200..<300` range.

The lines:
```java
      File file = new File(new URI(this.filePath));
      JSONObject fileEntry = FileUtils.getFilePlugin().getEntryForFile(file);
```
are the culprit somehow.

So far I've been unsuccessful at tracking down what the issue is. The second line comes from the docs [verbatim](https://github.com/apache/cordova-plugin-file/blob/master/doc/plugins.md#android) — where file creation is mentioned — but from a quick look at the code paths it hits I couldn't find a place where file / directory is created. The first line uses Java's own `File` and would not create anything.

My guess is that `cordova-plugin-file` dependency is failing to provide a valid file for the `BufferedOutputStream` (see above) to write to.

**Comment by jacobg** [comment]
Code push now broken on cordova-ios 6 due to use of file transfer plugin, as discussed here: #624. I can't believe Microsoft is not finishing up this pull request and merging it.

**Comment by ermik** [comment]
I am not with Microsoft, at least not until they buy me too, but if anyone can help based on my research into the remaining Android issue or guide me to a solution — I'm available.

**Comment by bobrosoft** [comment]
@matthiaswenz hi there! Can this one be merged asap? Causing bunch of issues with latest Cordova-ios 6 etc. https://github.com/microsoft/cordova-plugin-code-push/issues/624

**Comment by alexandergoncharov** [comment]
Hi all,
I'm sorry for this long delay.

I reproduced this issue https://github.com/microsoft/cordova-plugin-code-push/pull/513#issuecomment-613891164 and starting debugging to fix it.
Unfortunately, we can't merge this with the issues.
 @ermik , thanks for your findings! It is helpful. Also, could you please fix the conflict files?

**Comment by ermik** [comment]
Will fix conflicts.

On my end I am not yet using Cordova@6.0.0 so I don't think I will be able to fully take this on. But I'm available for specific questions / tasks.

**Comment by alexandergoncharov** [comment]
Small status:
It looks like some issues with paths. I can get update successfully if I use `cordova.file.dataDirectory` path(usually we use also 2 levels under that for keeping CodePush updates separately from other files). Trying also downloading this file in `cordova.file.dataDirectory` and moving to other but my attempts were unsuccessful. My work still in progress.

File: www/remotePackage.ts [hunk 3]
```diff
@@ -59,41 +66,30 @@ class RemotePackage extends Package implements IRemotePackage {
...
                 };
 
-                this.currentFileTransfer.onprogress = (progressEvent: ProgressEvent) => {
-                    if (downloadProgress) {
-                        var dp: DownloadProgress = { receivedBytes: progressEvent.loaded, totalBytes: progressEvent.total };
-                        downloadProgress(dp);
-                    }
-                };
+                const filedir = cordova.file.dataDirectory + LocalPackage.DownloadDir + "/";
+                const filename = LocalPackage.PackageUpdateFileName;
```
Comment (alexandergoncharov) [inline www/remotePackage.ts:74]:
```suggestion
                
                // Create a directory to download the update if it was not created
                FileUtil.getDataDirectory(LocalPackage.DownloadDir, true, (error: Error, dirEntry: DirectoryEntry) => {
                    if (error) {
                        CodePushUtil.logError("Can't create directory for download update.", error);
                    }
                });
                
```

**Comment by alexandergoncharov** [comment]
Hi all,
Looks like I found the solution: `cordova-plugin-advanced-http` plugin can't create a new directory for downloaded files and we should just create this one manually. I added a code suggestion. @ermik , thanks for fixing conflicts! Could you please also update the PR with this changes and test it? Also will be useful if someone else tests this PR fully?

It works for me on Android and with `cordova-ios@6.1.0`.

**Comment by ludufre** [comment]
> Hi all,
> Looks like I found the solution: `cordova-plugin-advanced-http` plugin can't create a new directory for downloaded files and we should just create this one manually. I added a code suggestion. @ermik , thanks for fixing conflicts! Could you please also update the PR with this changes and test it? Also will be useful if someone else tests this PR fully?
> 
> It works for me on Android and with `cordova-ios@6.1.0`.

Hi. I installed with `npm install microsoft/cordova-plugin-code-push#pull/513/head`. What next step to test with `cordova-ios@6.1.0`?

Thanks

**Comment by ermik** [comment]
@ludufre I think you would help by checking out this branch running `gulp test` on your machine, but if you want to use it in your code, you can simply lower the version of the app you use locally below you last CodePush release and then run the app on the simulator for iOS and Android. If you see CodePush succeed — then it's good news.

**Comment by ermik** [comment]
@alexandergoncharov I can't seem to be able to make `gulp` orchestrate the testing on my machine, I spent a bunch of time trying to make it see Cordova binary, but no luck. Opening this up for review and hopefully there will be one or two testers to verify this works with Cordova 5 & 6 alike.

**Comment by ludufre** [comment]
> @ludufre I think you would help by checking out this branch running `gulp test` on your machine, but if you want to use it in your code, you can simply lower the version of the app you use locally below you last CodePush release and then run the app on the simulator for iOS and Android. If you see CodePush succeed — then it's good news.

`gulp test` or `gulp test-ios` on main or this branch outputs: `Failed to restore plugin "cordova-plugin-whitelist" from config.xml. You might need to try adding it again. Error: Error: Cannot find module './src/superspawn'`
I strongly believe that I am doing something wrong.

**Comment by ludufre** [comment]
@ermik Just to be clear: with `cordova-ios@5.1.1` it works perfectly. Just with `cordova-ios@6.1.0` + this branch I have an error to compile, complains about UserAgent.

But regardless of that, I'm trying to help with the `gulp test` but it gives the error mentioned in the previous post.

**Comment by alexandergoncharov** [comment]
@ludufre, @ermik, Yeah, currently Cordova tests are broken(and before this PR). I also can run `gulp` command but can't run `gulp test` command. For now, we can test this changes only manually, unfortunately. 
I think that this broken tests will be a good item for the next work.

**Comment by ermik** [comment]
Do as you think is best, but I wouldn't merge any PR's until tests can be run. /inv @Krasavinigor @matthiaswenz

**Comment by alexandergoncharov** [comment]
Yeah, your thoughts are really good!
But I am just afraid that fixing tests can take really long time as it was broken for a really long time and this PR can unblock users that use new version of `cordova-ios`. So, it has big priority.
So, my suggestion here: let's try to fix tests for 2 days(until Tuesday) and if on that day we will not have running tests, let's merge it for unblock users and fix test in the future.

**Comment by ludufre** [comment]
Working iOS and Android with defaults platform versions.

```
Ionic:

   Ionic CLI                     : 6.10.1 (/Users/ludufre/.npm-packages/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 5.2.3
   @angular-devkit/build-angular : 0.901.11
   @angular-devkit/schematics    : 9.1.11
   @angular/cli                  : 9.1.11
   @ionic/angular-toolkit        : 2.2.0

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : android 8.1.0, ios 5.1.1
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.2.0, cordova-plugin-ionic-webview 4.2.1, (and 9 other plugins)

Utility:

   cordova-res : 0.15.1
   native-run  : 1.0.0

System:

   Android SDK Tools : 26.1.1 (/Users/ludufre/Library/Android/sdk)
   ios-deploy        : 1.10.0
   ios-sim           : 8.0.2
   NodeJS            : v12.18.1 (/usr/local/bin/node)
   npm               : 6.14.5
   OS                : macOS
   Xcode             : Xcode 11.6 Build version 11E708
```

With `cordova-ios@6.1.0` get error with deprecated `CDVUserAgentUtil`:

image

I didn't find any code that `[CDVUserAgentUtil ...acquireLock]` so I commented that line and compiled and worked perfectly in 6.1.1.

I made that small change, here it is: https://github.com/ludufre/cordova-plugin-code-push/commit/159e18cfa9e2aa0c5c6d47ca7b3f0e1d2b1c9f1e (I new on this Github thing, so I don't know how to send this change to this Pull Request / Or @ermik can do this change himself)

**Comment by ludufre** [comment]
Another thing that I observed, perhaps it's default plugin behavior:

If the Server Trust Mode is changed do 'pinned' (`this.http.setServerTrustMode('pinned')`) on the `cordova-plugin-advanced-http`, the CodePush stop working because SSL not trusted.

Because this settings is global and not per HTTP request maybe can't be circumvented.

We can add to README.md this behavior. What do you think?

(In my case I have to add the appcenter.ms cert to www/certificates because I use SSL Pinning)

**Comment by ermik** [comment]
@ludufre this is quite interesting, thank you for taking the time to poke around.

@alexandergoncharov I'm afraid I know nothing about gulp. This doesn't seem like a gulp-related issue, but I've had trouble navigating the test suites from the very beginning of this PR. 😭  I'm gonna try but please don't get your hopes up.

**Comment by andreidubov** [comment]
@alexandergoncharov @ermik @ludufre 

I started working on fixing tests [here](https://github.com/microsoft/cordova-plugin-code-push/tree/update-tests).

**Comment by alexandergoncharov** [comment]
Hi @ludufre ,
Thanks for debugging this PR :) 

Unfortunately, I can't reproduce your issue with `cordova-ios@6.1.1`. Also, as far as I know, the same for @andreidubov.
Could you please provide some demo app with the reproduced issue and reprosteps?

About Server Trust Mode: I think it is a good idea to add this info to docs. Could you please provide some code suggestions for this?

**Comment by szh** [comment]
I'm upgrading an app to use cordova-ios 6.1.0 and I'm having a compatibility issue with the file-transfer plugin. I'm using this fork until this PR is merged.

**Comment by szh** [comment]
I'm having the same issue as this user on Android with this fork.
https://github.com/microsoft/cordova-plugin-code-push/issues/611

**Comment by ludufre** [comment]
> Hi @ludufre ,
> Thanks for debugging this PR :)
> 
> Unfortunately, I can't reproduce your issue with `cordova-ios@6.1.1`. Also, as far as I know, the same for @andreidubov.
> Could you please provide some demo app with the reproduced issue and reprosteps?
> 
> About Server Trust Mode: I think it is a good idea to add this info to docs. Could you please provide some code suggestions for this?

Nevermind. I had cloned wrong PR. Fully working now `cordova-ios@6` and `cordova-android@9`.

> I'm having the same issue as this user on iOS with this fork.
> #611

The issue #611 have FileTransfer in the logs. Are you sure it's related? This fork don't use it.

**Comment by szh** [comment]
> The issue #611 have FileTransfer in the logs. Are you sure it's related? This fork don't use it.

Whether or not it's related - what's happening is that it updates successfully, and then it freezes on a white screen, and rolls back when I relaunch the application. I get the same critical log message:
```
[CodePush] Install succeeded.
=================================================================
Main Thread Checker: UI API called on a background thread: -[WKWebView loadFileURL:allowingReadAccessToURL:]
...
```

**Comment by ermik** [comment]
@szh, The unsafe `UIKit` call, in 90% of cases, would not cause a freeze or crash (some notable exceptions, no pun intended, are mutating calls to collections). It may also come from another Cordova dependency.

FWIW `UIKit` accessors, while a big no-no in the native world (and should be fixed, undoubtedly), have been unsafely called for a while, and many Cordova plugins are guilty of this.

**Comment by ermik** [comment]
@szh Thanks!

Please file your findings as an issue in this repo. It should be an easy fix whenever someone with decent Objective-C knowledge becomes available.

**Comment by alexandergoncharov** [comment]
Hi all,

We merged https://github.com/microsoft/cordova-plugin-code-push/pull/632 test fixing and verified this PR - all work correctly.
So, I'm going to merge this PR and make a new release.

@ermik, thanks a lot for your big work and contributing! It was valuable work! 👍

**Comment by jacobg** [comment]
Thank you very much!

Why does `plugin.xml` still have a dev dependency on `@types/cordova-plugin-file-transfer`?

**Comment by alexandergoncharov** [comment]
@jacobg Thanks! Good catch!

I'm removing it in this PR: https://github.com/microsoft/cordova-plugin-code-push/pull/633

**Comment by alexandergoncharov** [comment]
This PR already published in new version: https://github.com/microsoft/cordova-plugin-code-push/releases/tag/v1.13.0

**Comment by NickChowHK** [comment]
> This PR already published in new version: https://github.com/microsoft/cordova-plugin-code-push/releases/tag/v1.13.0

image

Should the version number be changed to 1.13.0?


Anyway...it still freezes on a blank screen after [CodePush] Install succeeded.
and rolls back when I relaunch the application


codePushSyncStatus 7
[CodePush] Downloading update
[CodePush] Package download success: {"deploymentKey":"xxx","label":"v1","appVersion":"1.0.0","isMandatory":false,"packageHash":"xxx","isFirstRun":false,"failedInstall":false,"localPath":"cdvfile://localhost/library-nosync/codepush/download/update.zip"}
codePushSyncStatus 8
[CodePush] Installing update
[CodePush] Applying full update
[CodePush] Install succeeded.
codePushSyncStatus 1

Main Thread Checker: UI API called on a background thread: -[WKWebView loadRequest:]
PID: 3311, TID: 1439918, Thread name: (none), Queue name: com.apple.root.default-qos, QoS: 0
Backtrace:
[CDVWebViewEngine loadRequest:] + 2080
[CodePush loadURL:] + 156
[CodePush loadPackage:] + 120
__20-[CodePush install:]_block_invoke + 652
libdispatch.dylib _dispatch_call_block_and_release + 24
libdispatch.dylib _dispatch_client_callout + 16
libdispatch.dylib _dispatch_queue_override_invoke + 872
ibdispatch.dylib _dispatch_root_queue_drain + 376
libdispatch.dylib _dispatch_worker_thread2 + 152
libsystem_pthread.dylib _pthread_wqthread + 212
libsystem_pthread.dylib start_wqthread + 8

[reports] Main Thread Checker: UI API called on a background thread: -[WKWebView loadRequest:]
PID: 3311, TID: 1439918, Thread name: (none), Queue name: com.apple.root.default-qos, QoS: 0
Backtrace:
-[CDVWebViewEngine loadRequest:] + 2080
-[CodePush loadURL:] + 156
-[CodePush loadPackage:] + 120
__20-[CodePush install:]_block_invoke + 652
libdispatch.dylib _dispatch_call_block_and_release + 24
libdispatch.dylib _dispatch_client_callout + 16
libdispatch.dylib _dispatch_queue_override_invoke + 872
libdispatch.dylib _dispatch_root_queue_drain + 376
libdispatch.dylib _dispatch_worker_thread2 + 152
libsystem_pthread.dylib _pthread_wqthread + 212
libsystem_pthread.dylib start_wqthread + 8

tcp_input [C4.1:3] flags=[R] seq=846688115, ack=0, win=0 state=LAST_ACK rcv_nxt=846688115, snd_una=2978486252
tcp_input [C4.1:3] flags=[R] seq=846688115, ack=0, win=0 state=CLOSED rcv_nxt=846688115, snd_una=2978486252
tcp_input [C4.1:3] flags=[R] seq=846688115, ack=0, win=0 state=CLOSED rcv_nxt=846688115, snd_una=2978486252
tcp_input [C5.1:3] flags=[R] seq=267997014, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=267997014, snd_una=3094212520
tcp_input [C6.1:3] flags=[R] seq=3760256249, ack=0, win=0 state=LAST_ACK rcv_nxt=3760256249, snd_una=1398964983
tcp_input [C6.1:3] flags=[R] seq=3760256249, ack=0, win=0 state=CLOSED rcv_nxt=3760256249, snd_una=1398964983
tcp_input [C6.1:3] flags=[R] seq=3760256249, ack=0, win=0 state=CLOSED rcv_nxt=3760256249, snd_una=1398964983

**Comment by ermik** [comment]
Sorry about missing `plugin.xml` but happy I can start my next PR. See you in 2 years!

**Pull request merged**